ta/pkcs11/src/pkcs11_attributes.c

```diff
@@ -2831,3 +2831,84 @@ enum pkcs11_rc add_missing_attribute_id(struct obj_attrs **pub_head,
 	else
 		return set_attribute(pub_head, PKCS11_CKA_ID, id2, id2_size);
 }
+
+enum pkcs11_rc check_attr_checksum(struct obj_attrs *head)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+
+	uint32_t value_len = 0;
+	uint32_t kcv_len = 0;
+	void *value = NULL;
+	void *kcv = NULL;
+
+	if (!head)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	/* CKA_CHECK_VALUE is also an attribute of CKO_CERTIFICATE */
+	if (get_class(head) != PKCS11_CKO_SECRET_KEY)
+		return PKCS11_RV_NOT_IMPLEMENTED;
+
+	if (get_key_type(head) != PKCS11_CKK_AES)
+		return PKCS11_RV_NOT_IMPLEMENTED;
+
+	/* Get key CKA_VALUE */
+	rc = get_attribute_ptr(head, PKCS11_CKA_VALUE, &value, &value_len);
+	if (rc)
+		return rc;
+
+	/* Get key check value size */
+	rc = aes_get_check_value_size(value, value_len, &kcv_len);
+	if (rc)
+		goto out;
+
+	kcv = TEE_Malloc(kcv_len, TEE_MALLOC_FILL_ZERO);
+	if (!kcv) {
+		rc =  PKCS11_CKR_DEVICE_MEMORY;
+		goto out;
+	}
+
+	/* Compute and add the key check sum value */
+	rc = aes_compute_check_value(value, value_len, kcv, &kcv_len);
+	if (rc)
+		goto out;
+
+	/* Check whether CKA_CHECK_VALUE has been provided in the template */
+	rc = get_attribute_ptr(head, PKCS11_CKA_CHECK_VALUE,
+			       &value, &value_len);
+
+	if (rc != PKCS11_CKR_OK && rc != PKCS11_RV_NOT_FOUND) {
+		rc = PKCS11_CKR_GENERAL_ERROR;
+		goto out;
+	}
+
+	/*
+	 * The generation of the KCV may be prevented by the application
+	 * supplying the attribute in the template as a no-value (0 length)
+	 * entry.
+	 */
+	if (rc == PKCS11_CKR_OK) {
+		if (value_len == 0)
+			goto out;
```

Comment:
this test could be done before computation of the value: not needed if client does not want the attribute to be set.